src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -567,13 +569,56 @@ private function writeProperty(&$object, $property, $value)
 
             $object->$property = $value;
         } elseif (self::ACCESS_TYPE_MAGIC === $access[self::ACCESS_TYPE]) {
-            $object->{$access[self::ACCESS_NAME]}($value);
+            $this->callMethod($object, $access[self::ACCESS_NAME], $value);
         } else {
             throw new NoSuchPropertyException($access[self::ACCESS_NAME]);
         }
     }
 
     /**
+     * Call a method and convert {@see \TypeError} to {@see InvalidArgumentException}.
+     *
+     * @param object $object
+     * @param string $method
+     * @param mixed  $value
+     *
+     * @throws InvalidArgumentException
+     * @throws \Exception
+     */
+    private function callMethod($object, $method, $value) {
+        // Cannot use class_exists because the PHP 7 polyfill defines \TypeError
+        if (PHP_MAJOR_VERSION >= 7) {
+            // PHP 7
+            try {
+                $object->{$method}($value);
+            } catch (\TypeError $e) {
+                throw new InvalidArgumentException($e->getMessage(), $e->getCode(), $e);
+            }
+
+            return;
+        }
+
+        // PHP 5
+        set_error_handler(function ($errno, $errstr) {
```

Comment:
I prefer to not test the error message. It's not clean and it still does not handle all cases. Throwing an error (and not an exception) in a mutator should be very very rare in a real code.